### PR TITLE
Perf actual React components

### PR DESCRIPTION
**Pull request by gfogle**
Working through #5254 

I'm hijacking the lifecycle methods of the react components and then enabling DefaultPerf to measure them and print high-level statistics on the measures.

I started at a basic implementation. Would like to submit subsequent PRs with more advanced functionality such as tracking # of times returned true vs false for shouldComponentUpdate, etc.

### CODE REVIEW

**Comment by gfogle** [comment]
FYI - travis is complaining about dangling comma rule. I get that this simplifies the diff, but it really looks weird to see those. fixing now though.

File: src/renderers/shared/reconciler/ReactCompositeComponent.js [hunk 1]
```diff
@@ -262,6 +262,28 @@ var ReactCompositeComponentMixin = {
...
+
+      perfMethods.forEach(function(method) {
+        if (this._instance[method] && typeof this._instance[method] === 'function') {
+          this._instance[method] = ReactPerf.measure(
+            'ReactCompositeComponent',
+            method,
+            this._instance[method]
+          );
+        }
+      }.bind(this));
```
Comment (yaycmyk) [inline src/renderers/shared/reconciler/ReactCompositeComponent.js]: `bind` isn't necessary, `forEach` takes a second arg, being `thisarg` if desired

File: src/test/ReactDefaultPerf.js [hunk 2]
```diff
@@ -96,6 +96,26 @@ var ReactDefaultPerf = {
...
+  getLifecyclesSummaryMap: function(measurements) {
+    var summary = ReactDefaultPerfAnalysis.getLifecycleSummary(
+      measurements
+    );
+    return summary.map(function(item) {
+      var flattenedItem = {
+        id: item.id,
+      };
+      var lifecycles = Object.keys(item.measures);
+      lifecycles.forEach(function(method) {
```
Comment (yaycmyk) [inline src/test/ReactDefaultPerf.js]: `reduce` might make more sense here

File: src/test/ReactDefaultPerf.js [hunk 3]
```diff
@@ -121,6 +141,29 @@ var ReactDefaultPerf = {
...
+  printLifecycles: function(measurements) {
+    measurements = measurements || ReactDefaultPerf._allMeasurements[
+      ReactDefaultPerf._allMeasurements.length - 1
+    ];
+    var components = measurements.components;
+    var summary = ReactDefaultPerfAnalysis.getLifecycleSummary(components);
+    console.table(summary.map(function(item) {
+      var result = {};
+      var keys = Object.keys(item);
+      keys.forEach(function(key) {
```
Comment (yaycmyk) [inline src/test/ReactDefaultPerf.js]: same here, `reduce` would probably make more sense

Comment (gfogle) [inline src/test/ReactDefaultPerf.js]: i kept this implementation consistent with the 'getMeasurementsSummaryMap' method above

File: src/test/ReactDefaultPerf.js (hunk 2 above)
Comment (gfogle) [inline src/test/ReactDefaultPerf.js]: see comment below

File: src/renderers/shared/reconciler/ReactCompositeComponent.js (hunk 1 above)
Comment (gfogle) [inline src/renderers/shared/reconciler/ReactCompositeComponent.js]: good point

**Comment by gfogle** [comment]
@sebmarkbage - I think this is a much clearer picture of the prior PR i closed because the commit history was just not worth rebasing out to reflect final implementation.

File: src/test/ReactDefaultPerf.js (hunk 3 above)
Comment (vramana) [inline src/test/ReactDefaultPerf.js]: @gfogle You are converting a collection (`keys`) into single object (`result`). So it is essentially a reduce.

Comment (gfogle) [inline src/test/ReactDefaultPerf.js]: thats a fair point. if whoever reviews the PR wants me to update it to use reduce i'm happy to do so. not clear if they would like to keep consistent with the other print methods already in the file.

**Comment by sebmarkbage** [comment]
We've just recently started work on a new perf tool. #6046 

This will allow us to do refactoring to the code base without breaking the perf tools all the time. Could you maybe check if your issue can be solved in #6046? Sorry for the inconvenience. We didn't plan for this new tool to replace the old one at this point.

**Pull request closed without merging**